### PR TITLE
Try to make EDD support actually recognize things other than by the MBR matching...

### PR DESCRIPTION
**Pull request by vathpela**
Basically the parser for ATA will kind of work with SCSI if lun==0, but
it won't work with SATA or ATA at all, and the parser we had for SCSI
will only work with virtio.

So in the past we nearly always match things based on MBR signature,
which is pretty awful.

This patch separates those all out by which kernel driver is actually
providing the sysfs devices, and tries to match the EDD fields up with
how they're actually provided by the kernel.

Signed-off-by: Peter Jones <pjones@redhat.com>

### CODE REVIEW

**Comment by vathpela** [comment]
FWIW this also changes the log to give us enough information to maybe know /why/ it may or may not have worked:

19:54:19,378 INFO blivet: edd: MBR signature on sda is zero. new disk image?
19:54:19,386 INFO blivet: edd: MBR signature on sdc is zero. new disk image?
19:54:19,387 INFO blivet: edd: collected mbr signatures: {'sdb': '0x00033091', 'sdd': '0x64462d2c'}
19:54:19,387 WARN blivet: edd: interface type USB is not implemented (/sys/firmware/edd/int13_dev80)
19:54:19,388 DEBUG blivet: edd: data extracted from 0x80:
	path: /sys/firmware/edd/int13_dev80 version: 0x21
	sysfs pci path: 
	type: USB, ata_device: None
	channel: 255, mbr_signature: 0x64462d2c
	pci_dev: ff:ff.255, scsi_id: None, scsi_lun: None, sectors: 15667200
19:54:19,389 INFO blivet: edd: matched 0x80 to sdd using MBR sig
19:54:19,390 DEBUG blivet: edd: data extracted from 0x81:
	path: /sys/firmware/edd/int13_dev81 version: 0x30
	sysfs pci path: ../devices/pci0000:00/0000:00:1f.2/ata3/host2/target2:0:0/2:0:0:0/block/sdb
	type: SATA, ata_device: 2
	channel: 255, mbr_signature: 0x00033091
	pci_dev: 00:1f.2, scsi_id: None, scsi_lun: None, sectors: 312581808
19:54:19,390 INFO blivet: edd: matched 0x81 to sdb using pci_dev
19:54:19,391 DEBUG blivet: edd: data extracted from 0x82:
	path: /sys/firmware/edd/int13_dev82 version: 0x30
	sysfs pci path: ../devices/pci0000:00/0000:00:1f.2/ata1/host0/target0:0:0/0:0:0:0/block/sda
	type: SATA, ata_device: 0
	channel: 255, mbr_signature: 0x00000000
	pci_dev: 00:1f.2, scsi_id: None, scsi_lun: None, sectors: 312581808
19:54:19,392 INFO blivet: edd: matched 0x82 to sda using pci_dev
19:54:19,393 DEBUG blivet:          PartitionDevice.setup: sdd1 ; controllable: True ; orig: False ; status: True ;
19:54:19,407 DEBUG blivet:           Ext4FS.setup: device: /dev/sdd1 ; type: ext4 ; status: True ;

**Comment by bcl** [comment]
Looks good to me.

File: blivet/devicelibs/edd.py [hunk 1]
```diff
@@ -32,66 +32,193 @@
...
+
+        # data from edd/int13_devXX/interface when type is I2O
+        self.i2o_identity = None
+
+        # data from edd/int13_devXX/interface when type is SAS
+        self.sas_address = None
+        self.sas_lun = None
+
+        # data we may get as the result of the search
+        self.sysfslink = None
```
Comment (vpodzime) [inline blivet/devicelibs/edd.py]: All these descriptions in comments could be turned into an actual documentation by changing the comments to be docstrings (`"""`strings right below the attributes)

File: blivet/devicelibs/edd.py [hunk 2]
```diff
@@ -32,66 +32,193 @@
...
-            "\ttype: %(type)s, ata_device: %(ata_device)s\n" \
-            "\tchannel: %(channel)s, mbr_signature: %(mbr_signature)s\n" \
-            "\tpci_dev: %(pci_dev)s, scsi_id: %(scsi_id)s\n" \
-            "\tscsi_lun: %(scsi_lun)s, sectors: %(sectors)s" % self.__dict__
+        s = "\tpath: %(sysfspath)s, version: %(version)s\n" \
+            "\ttype: %(type)s, mbr_signature: %(mbr_sig)s, " \
+            "sectors: %(sectors)s"
+        if self.sysfslink != None:
+            s += "\n\tsysfs pci path: %(sysfslink)s"
+        if self.host_bus !=None or self.pci_dev != None or self.channel != None:
```
Comment (vpodzime) [inline blivet/devicelibs/edd.py]: Missing space after the first ``!=``

Comment (vpodzime) [inline blivet/devicelibs/edd.py]: Also, I think
```
if any(self.host_bus, self.pci_dev, self.channel):
```
would be better readable.

**Comment by vpodzime** [comment]
Otherwise this looks good to me too. It would be nice to have some tests for this code. We could probably create our own artificial ``/sys`` tree somewhere and then ``chroot()`` right above it to get a testing environment.

**Comment by dwlehman** [comment]
I'll second the call for unit tests. I'm pretty sure the mock module provides a way to mock filesystem trees.

**Comment by clumens** [comment]
It looks like your first commit removed the only call to open() in edd.py, which is what pylint is complaining about.  Remove that, and ACK from me.

**Pull request merged**